synapse/storage/databases/state/bg_updates.py

```diff
@@ -173,7 +186,8 @@ def _get_state_groups_from_groups_txn(
                         allow_none=True,
                     )
 
-        return results
+        # The results shouldn't be considered mutable.
+        return cast(Dict[int, StateMap[str]], results)
```

Comment:
This is pretty dirty, but I couldn't figure out a way for it to be mutable in this method, but not mutable on return...

Reply:
I typed this function in one of my abandoned PRs and don't recall needing any casts.

`MutableStateMap[X]` is a subtype of `StateMap[X]`, so this will work as long as you say this is the return type: `Mapping[int, StateMap[str]]` (iirc).

Reply:
That's what i had originally and it didn't like it.

Reply:
> That's what i had originally and it didn't like it.

Why not?

Reply:
Oh I thought you said **you** didn't like it. I tried this just now on a copy of your branch and it worked fine

Reply:
Ah, the trick seems to be to use `Mapping`, not `Dict` as the outer container. (o_O)

Reply:
yes, because `Dict` can't be covariant in its value type (due to its mutability)